### DISCUSSION TITLE
I would like to incorporate a new security check OSS tool into the starboard.

### DISCUSSION THREAD

**Discussion by hiropinponpan** [aquasecurity/starboard#300] [Development]
Right now, I think there are four tools, kubebench, configaudit, kubehunter, and vulns, but I would like to incorporate the existing popeye (A Kubernetes Cluster Sanitizer).
I think it is important to detect and deal with unused resources of k8s cluster in order to reduce unnecessary attack points of k8s cluster.
What do you think of this idea?

**Comment by danielpacak** [comment]
👋 @hiropinponpan IMO That's great idea! Our assumption was that you could add and swap config auditors or vulnerability scanners, but at the same time reuse infrastructure (i.e. Kubernetes jobs scheduler and custom resources) provided by Starboard toolkit. As an example, check how you can swap static vulnerability scanner by changing Starboard's config https://aquasecurity.github.io/starboard/integrations/vulnerability-scanners/

I know popeye and I think it's great tool to integrate with Starboard as an addition to Polaris or as a new type of scanner with its own report schema (= custom resource definition).

To get started I'd check the following:
- [ ] Can you run popeye as a Kubernetes job?
- [ ] Can you run popeye for a single Kubernetes workload? (That was one feature that we requested Polaris to add. Initially they scanned all workloads in the cluster.)
- [ ] Check if you can integrate with the current command-line interface, e.g. add the `--scanner` flag to choose Polaris vs popeye: 
  ```
   starboard generate configauditreport deployment/nginx --scanner=polaris
   starboard generate configauditreport deployment/nginx --scanner=popeye
   ```
- [ ] Get familiar with ConfigAudiReport schema and see if popeye results fit into that. If not, think about extending ConfigAuditReport to make it more generic and suitable for both Polaris and popeye. Ideally ConfigAuditReport should work with other configuration auditors / sanitizers / linters

Based on this preliminary check list we can think about next steps. /cc @lizrice

**Reply by hiropinponpan** [reply]
@danielpacak  Thanks for your comment. 

- [ ] It has been confirmed that popeye can be run with k8s job and reported. However, if there is some issue in the popeye check result item, the exit code will be 1, so the job status will look like fail.  I think it is necessary to consider this exit code when calling popeye from starboard-cli.
- [ ] popeye can check only specific k8s resources and namespaces with execution arguments and config files (spinach). I think this feature can specify and check a specific k8s workload.
- [ ] I think popeye has a different checking purpose than other tools (polaris), so I thought it would be better to add a new set from crd.　I actually added and tried it.　 
```
$ k get crd popeyereports.aquasecurity.github.io
NAME                                   CREATED AT
popeyereports.aquasecurity.github.io   2021-01-09T08:56:05Z

$ starboard scan popeyereports -v 5
I0109 18:10:13.139264    6608 runner.go:79] Running task and waiting forever
I0109 18:10:13.141152    6608 runnable_job.go:72] Creating job "starboard/8096476e-8945-4bd4-bc88-639fd0384af6"
I0109 18:10:13.176842    6608 reflector.go:175] Starting reflector *v1.Job (30m0s) from pkg/mod/k8s.io/client-go@v0.18.6/tools/cache/reflector.go:125
I0109 18:10:13.176904    6608 reflector.go:211] Listing and watching *v1.Job from pkg/mod/k8s.io/client-go@v0.18.6/tools/cache/reflector.go:125
I0109 18:10:13.279077    6608 shared_informer.go:253] caches populated
I0109 18:10:16.893455    6608 runnable_job.go:107] Stopping runnable job on task completion with status: Complete
I0109 18:10:16.893538    6608 runner.go:83] Stopping runner on task completion with error: <nil>
I0109 18:10:16.893552    6608 scanner.go:80] Getting logs for popeye container in job: starboard/8096476e-8945-4bd4-bc88-639fd0384af6
I0109 18:10:17.416269    6608 scanner.go:72] Deleting job: starboard/8096476e-8945-4bd4-bc88-639fd0384af6

$ k get popeyereport.aquasecurity.github.io/cluster -o yaml | head -20
apiVersion: aquasecurity.github.io/v1alpha1
kind: PopeyeReport
metadata:
  creationTimestamp: "2021-01-09T09:10:17Z"
  generation: 1
  labels:
    starboard.resource.kind: Cluster
    starboard.resource.name: cluster
  managedFields:
  - apiVersion: aquasecurity.github.io/v1alpha1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:labels:
          .: {}
          f:starboard.resource.kind: {}
          f:starboard.resource.name: {}
      f:report:
        .: {}
        f:popeye:
```

But should I consider it as part of the configauditreport? what do you think？

**Reply by danielpacak** [reply]
Nice progress @hiropinponpan Let me play with Popeye and I'll get back to you with my opinion whether we should reuse/extend configauditreports resource or come up with a new one.

**Reply by danielpacak** [reply]
[@hiropinponpan](url) I think the main idea behind Starboard was to standardise and host a minimal set of different reports or, to the extreme have a single CRD to represent all reports. This is for example what Kubernetes [wg-policy](https://github.com/kubernetes-sigs/wg-policy-prototypes/blob/master/policy-report/README.md) is trying to pursue and we'll probably join forces with them.

What is more we always wanted to leverage Kubernetes API machinery to let a security report inhering a life cycle of the underlying Kubernetes object. For example, delete a ConfigAuditReport whenever the owning ReplicaSet is deleted. This currently (v0.10) only works for Kubernetes controllers (e.g. naked Pods, ReplicaSets, DaemonSets, etc.), but we plan to extend it to other kinds such as Services or RBAC configuration objects.

Having that said, IMO Popey falls under the same category of configuration checkers as Polaris and Conftest so instead of defining a tool-specific CRD called PopeyeReport I'd evaluate whether it's possible to implement the [`configaduitreport.Plugin`](https://pkg.go.dev/github.com/aquasecurity/starboard@v0.10.1/pkg/configauditreport#Plugin) for Popeye and transform the output of Popeye to ConfigAuditReport.

Notice also that this would require making the ConfigAuditReport more generic to cater for all kinds of Kubernetes object and probably introducing ClusterConfigAuditReport to associate it with cluster-scoped object but this is anyhow on our roadmap with near-term priority. /cc @itaysk
